// moltbook/dist/index.mjs
//#region src/errors.ts
var MoltbookError = class extends Error {
	constructor(args) {
		super(args.message);
		this.name = "MoltbookError";
		this.status = args.status;
		this.statusText = args.statusText;
		this.method = args.method;
		this.url = args.url;
		this.body = args.body;
	}
};

//#endregion
//#region src/moltbook.ts
function toQueryString(query) {
	if (!query) return "";
	const params = new URLSearchParams();
	for (const [k, v] of Object.entries(query)) {
		if (v === void 0 || v === null) continue;
		params.set(k, String(v));
	}
	const s = params.toString();
	return s ? `?${s}` : "";
}
function isBlob(value) {
	return typeof Blob !== "undefined" && value instanceof Blob;
}
function toArrayBufferCopy(u8) {
	const ab = new ArrayBuffer(u8.byteLength);
	new Uint8Array(ab).set(u8);
	return ab;
}
function toBlob(data, contentType) {
	if (typeof Blob === "undefined") throw new Error("Blob is not available in this runtime. Use Node 18+ / modern browsers, or pass a Blob directly.");
	const blobPart = data instanceof Uint8Array ? toArrayBufferCopy(data) : data;
	return new Blob([blobPart], contentType ? { type: contentType } : void 0);
}
function toFormData(file) {
	if (typeof FormData === "undefined") throw new Error("FormData is not available in this runtime. Use Node 18+ / modern browsers, or provide a fetch/FormData polyfill.");
	const fd = new FormData();
	if (isBlob(file)) {
		fd.append("file", file, "avatar");
		return fd;
	}
	const blob = isBlob(file.data) ? file.data : toBlob(file.data, file.contentType);
	fd.append("file", blob, file.filename);
	return fd;
}
var Moltbook = class {
	constructor(opts) {
		this.apiKey = opts.apiKey;
		this.baseUrl = "https://www.moltbook.com/api/v1";
		this.fetchImpl = globalThis.fetch;
		if (!this.fetchImpl) throw new Error("No fetch implementation found. Use a runtime that provides global fetch (Node 18+ / modern browsers).");
	}
	async request(args) {
		const url = `${this.baseUrl}${args.path}${toQueryString(args.query)}`;
		const headers = new Headers();
		headers.set("Authorization", `Bearer ${this.apiKey}`);
		let body = void 0;
		if (args.formData) body = args.formData;
		else if (args.body !== void 0) {
			body = JSON.stringify(args.body);
			if (!headers.has("Content-Type")) headers.set("Content-Type", "application/json");
		}
		const res = await this.fetchImpl(url, {
			method: args.method,
			headers,
			body
		});
		if (res.status === 204) return void 0;
		const parsedBody = (res.headers.get("content-type") ?? "").includes("application/json") ? await res.json().catch(() => void 0) : await res.text().catch(() => void 0);
		if (!res.ok) throw new MoltbookError({
			message: `Moltbook API error ${res.status} ${res.statusText}`,
			status: res.status,
			statusText: res.statusText,
			method: args.method,
			url,
			body: parsedBody
		});
		return parsedBody;
	}
	getMe() {
		return this.getProfile();
	}
	getProfile() {
		return this.request({
			method: "GET",
			path: "/agents/me"
		});
	}
	getClaimStatus() {
		return this.request({
			method: "GET",
			path: "/agents/status"
		});
	}
	getAgentProfile(name) {
		return this.request({
			method: "GET",
			path: "/agents/profile",
			query: { name }
		});
	}
	updateProfile(body) {
		return this.request({
			method: "PATCH",
			path: "/agents/me",
			body
		});
	}
	uploadAvatar(file) {
		return this.request({
			method: "POST",
			path: "/agents/me/avatar",
			formData: toFormData(file)
		});
	}
	removeAvatar() {
		return this.request({
			method: "DELETE",
			path: "/agents/me/avatar"
		});
	}
	followAgent(name) {
		return this.request({
			method: "POST",
			path: `/agents/${encodeURIComponent(name)}/follow`
		});
	}
	unfollowAgent(name) {
		return this.request({
			method: "DELETE",
			path: `/agents/${encodeURIComponent(name)}/follow`
		});
	}
	createTextPost(body) {
		return this.createPost(body);
	}
	createLinkPost(body) {
		return this.createPost(body);
	}
	createPost(body) {
		return this.request({
			method: "POST",
			path: "/posts",
			body
		});
	}
	getPosts(params) {
		return this.request({
			method: "GET",
			path: "/posts",
			query: params ? {
				sort: params.sort,
				limit: params.limit,
				submolt: params.submolt
			} : void 0
		});
	}
	getSubmoltFeed(submoltName, params) {
		return this.request({
			method: "GET",
			path: `/submolts/${encodeURIComponent(submoltName)}/feed`,
			query: params ? {
				sort: params.sort,
				limit: params.limit
			} : void 0
		});
	}
	getPost(postId) {
		return this.request({
			method: "GET",
			path: `/posts/${encodeURIComponent(postId)}`
		});
	}
	deletePost(postId) {
		return this.request({
			method: "DELETE",
			path: `/posts/${encodeURIComponent(postId)}`
		});
	}
	addComment(postId, body) {
		return this.request({
			method: "POST",
			path: `/posts/${encodeURIComponent(postId)}/comments`,
			body
		});
	}
	getComments(postId, params) {
		return this.request({
			method: "GET",
			path: `/posts/${encodeURIComponent(postId)}/comments`,
			query: params ? { sort: params.sort } : void 0
		});
	}
	upvotePost(postId) {
		return this.request({
			method: "POST",
			path: `/posts/${encodeURIComponent(postId)}/upvote`
		});
	}
	downvotePost(postId) {
		return this.request({
			method: "POST",
			path: `/posts/${encodeURIComponent(postId)}/downvote`
		});
	}
	upvoteComment(commentId) {
		return this.request({
			method: "POST",
			path: `/comments/${encodeURIComponent(commentId)}/upvote`
		});
	}
	createSubmolt(body) {
		return this.request({
			method: "POST",
			path: "/submolts",
			body
		});
	}
	listSubmolts() {
		return this.request({
			method: "GET",
			path: "/submolts"
		});
	}
	getSubmolt(name) {
		return this.request({
			method: "GET",
			path: `/submolts/${encodeURIComponent(name)}`
		});
	}
	subscribeToSubmolt(name) {
		return this.request({
			method: "POST",
			path: `/submolts/${encodeURIComponent(name)}/subscribe`
		});
	}
	unsubscribeFromSubmolt(name) {
		return this.request({
			method: "DELETE",
			path: `/submolts/${encodeURIComponent(name)}/subscribe`
		});
	}
	getFeed(params) {
		return this.request({
			method: "GET",
			path: "/feed",
			query: params ? {
				sort: params.sort,
				limit: params.limit
			} : void 0
		});
	}
	search(params) {
		return this.request({
			method: "GET",
			path: "/search",
			query: {
				q: params.q,
				type: params.type,
				limit: params.limit
			}
		});
	}
};

//#endregion
export { Moltbook, MoltbookError };
